docs/source/tutorial/convnet.rst

```diff
@@ -0,0 +1,428 @@
+            super(ResNet152, self).__init__(
+                conv1=L.Convolution2D(
+                    None, 64, 7, 2, 3, initialW=w, nobias=True),
+                bn1=L.BatchNormalization(64),
+                res2=ResBlock(n_blocks[0], 64, 64, 256, 1),
+                res3=ResBlock(n_blocks[1], 256, 128, 512),
+                res4=ResBlock(n_blocks[2], 512, 256, 1024),
+                res5=ResBlock(n_blocks[3], 1024, 512, 2048))
+            self.train = True
+
+        def __call__(self, x):
+            h = self.bn1(self.conv1(x), test=not self.train)
+            h = F.max_pooling_2d(F.relu(h), 2, 2)
+            h = self.res2(h)
+            h = self.res3(h)
+            h = self.res4(h)
+            h = self.res5(h)
+            if self.train:
+                return h
+            return F.softmax(h)
+
+    class ResBlock(chainer.ChainList):
+
+        def __init__(self, n_layers, n_in, n_mid, n_out, stride=2):
+            w = chainer.initializers.HeNormal()
+            super(ResBlock, self).__init__()
+            self.add_link(BottleNeck(n_in, n_mid, n_out, stride, True))
+            for _ in range(n_layers - 1):
+                self.add_link(BottleNeck(n_out, n_mid, n_out))
+
+        def __call__(self, x, train):
+            for f in self.children():
+                x = f(x, train)
+            return x
+
+    class BottleNeck(chainer.Chain):
+
+        def __init__(self, n_in, n_mid, n_out, stride=1, proj=False):
+            w = chainer.initializers.HeNormal()
+            super(BottleNeck, self).__init__(
+                conv1x1a=L.Convolution2D(
+                    n_in, n_mid, 1, stride, 0, initialW=w, nobias=True),
+                conv3x3b=L.Convolution2D(
+                    n_mid, n_mid, 3, 1, 1, initialW=w, nobias=True),
+                conv1x1c=L.Convolution2D(
+                    n_mid, n_out, 1, 1, 0, initialW=w, nobias=True),
+                bn_a=L.BatchNormalization(n_mid),
+                bn_b=L.BatchNormalization(n_mid),
+                bn_c=L.BatchNormalization(n_out))
+            if proj:
+                self.add_link('conv1x1r', L.Convolution2D(
+                    n_in, n_out, 1, stride, 0, initialW=w, nobias=True))
+                self.add_link('bn_r', L.BatchNormalization(n_out))
+            self.proj = proj
+
+        def __call__(self, x, train):
+            h = F.relu(self.bn_a(self.conv1x1a(x), test=not train))
+            h = F.relu(self.bn_b(self.conv3x3b(x), test=not train))
+            h = self.bn_c(self.conv1x1c(x), test=not train)
+            if self.proj:
+                x = self.bn_r(self.conv1x1r(x), test=not train)
+            return F.relu(h + x)
+
+In :class:`BottleNeck` class, it switches whther ``conv1x1r`` to fit the
+channel dimension of the input ``x`` and the output ``h`` and ``bn_r`` for it
+should be added or not with respect to :attr:`~BottleNeck.proj` attribute.
+Writing the building block in this way gains the re-usability of a class a lot.
```

Comment:
How about:

it will conditionally compute a convolutional layer conv1x1r which will extend the number of channels of the input x to be equal to the number of channels of the output of conv1x1c, and followed by a batch normalization layer before the final ReLU layer.